clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -381,21 +391,28 @@ public void commitAsync(OffsetCommitCallback callback) {
     @Override
     public void commitAsync(Map<TopicPartition, OffsetAndMetadata> offsets, OffsetCommitCallback callback) {
         CompletableFuture<Void> future = commit(offsets, false);
-        final OffsetCommitCallback commitCallback = callback == null ? new DefaultOffsetCommitCallback() : callback;
         future.whenComplete((r, t) -> {
-            if (t != null) {
-                commitCallback.onComplete(offsets, new KafkaException(t));
-            } else {
-                commitCallback.onComplete(offsets, null);
+            if (callback == null) {
+                if (t != null) {
+                    log.error("Offset commit with offsets {} failed", offsets, t);
+                }
+                return;
             }
-        }).exceptionally(e -> {
-            throw new KafkaException(e);
+
+            invoker.submit(new OffsetCommitCallbackTask(callback, offsets, (Exception) t));
         });
     }
 
     // Visible for testing
     CompletableFuture<Void> commit(Map<TopicPartition, OffsetAndMetadata> offsets, final boolean isWakeupable) {
+        maybeThrowFencedInstanceException();
+        maybeInvokeCallbacks();
         maybeThrowInvalidGroupIdException();
+
+        if (offsets.isEmpty()) {
+            return CompletableFuture.completedFuture(null);
+        }
```

Comment:
Maybe not related to this concrete PR, but the original commit code has some logic around updating the last seen leader epochs here, can we omit it, or is it a follow-up task?

Reply:
Thanks, I'll add a PR for investigation. I think metadata was not in the implementation at the time the function was implemented.

Reply:
Here is the Jira: https://issues.apache.org/jira/browse/KAFKA-15803